components/base/sensorcontrolled/movestraight.go

```diff
@@ -0,0 +1,204 @@
+		sb.logger.CWarnf(ctx,
+			"Position reporting sensor not available, and no control loop is configured, using base %s MoveStraight",
+			sb.controlledBase.Name().ShortName())
+		sb.stopLoop()
+		return sb.controlledBase.MoveStraight(ctx, distanceMm, mmPerSec, extra)
+	}
+
+	initialHeading, err := sb.headingFunc(ctx)
+	if err != nil {
+		return err
+	}
+	// make sure the control loop is enabled
+	if sb.loop == nil {
+		if err := sb.startControlLoop(); err != nil {
+			return err
+		}
+	}
+
+	// initialize relevant parameters for moving straight
+	slowDownDist := calcSlowDownDist(distanceMm)
+
+	var initPos *geo.Point
+
+	if sb.position != nil {
+		initPos, _, err = sb.position.Position(ctx, nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	ticker := time.NewTicker(time.Duration(1000./sb.controlLoopConfig.Frequency) * time.Millisecond)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			// do not return context canceled errors, just log them
+			if errors.Is(ctx.Err(), context.Canceled) {
+				sb.logger.Error(ctx.Err())
+				return nil
+			}
+			return ctx.Err()
+		case <-ticker.C:
+			var errDist float64
+
+			angVelDes, err := sb.calcHeadingControl(ctx, initialHeading)
+			if err != nil {
+				return err
+			}
+
+			if sb.position != nil {
+				errDist, err = sb.calcPositionError(ctx, distanceMm, initPos)
+				if err != nil {
+					return err
+				}
+			}
+
+			if errDist < moveStraightErrTarget {
+				return sb.Stop(ctx, nil)
+			}
+
+			linVelDes := calcLinVel(errDist, mmPerSec, slowDownDist)
+			if err != nil {
+				return err
+			}
+
+			// update velocity controller
+			if err := sb.updateControlConfig(ctx, linVelDes/1000.0, angVelDes); err != nil {
+				return err
+			}
+
+			// exit if the straight takes too long
+			if time.Since(startTime) > timeOut {
+				sb.logger.CWarn(ctx, "exceeded time for MoveStraightCall, stopping base")
+
+				return sb.Stop(ctx, nil)
+			}
+		}
+	}
+}
+
+// calculate the desired angular velocity to correct the heading of the base.
+func (sb *sensorBase) calcHeadingControl(ctx context.Context, initHeading float64) (float64, error) {
+	currHeading, err := sb.headingFunc(ctx)
+	if err != nil {
+		return 0, err
+	}
+
+	headingErr := initHeading - currHeading
+	headingErrWrapped := headingErr - (math.Floor((headingErr+180.)/(2*180.)))*2*180. // [-180;180)
+
+	return headingErrWrapped * headingGain, nil
+}
+
+// calcPositionError calculates the current error in position.
+// This results in the distance the base needs to travel to reach the goal.
+func (sb *sensorBase) calcPositionError(ctx context.Context, distanceMm int, initPos *geo.Point) (float64, error) {
+	pos, _, err := sb.position.Position(ctx, nil)
+	if err != nil {
+		return 0, err
+	}
+
+	// the currDist will always return as positive, so we need the goal distanceMm to be positive
+	currDist := initPos.GreatCircleDistance(pos) * 1000000.
+	return math.Abs(float64(distanceMm)) - currDist, nil
```

Comment:
the distance from the `GreatCircleDistance` calculation always be positive, so I needed the setpoint for the position term to also be positive. negative distances get handled in a different spot

Reply:
(maybe it will become more obvious as I get farther into this review, but) is it possible to handle negative distances all in this function as well just so it's all in the same spot?